File: app/src/main/java/com/app/ralaunch/controls/editor/ControlEditorOperations.java
package com.app.ralaunch.controls.editor;

import android.app.AlertDialog;
import android.content.Context;
import android.util.DisplayMetrics;
import android.widget.Toast;

import com.app.ralaunch.R;
import com.app.ralaunch.controls.ControlConfig;
import com.app.ralaunch.controls.ControlData;
import com.app.ralaunch.controls.ControlDataConverter;
import com.app.ralaunch.model.ControlElement;
import com.app.ralaunch.utils.ControlLayoutManager;
import com.app.ralaunch.utils.AppLogger;

/**
 * 控件编辑器操作类
 * 
 * 统一管理控件编辑器的共同操作，包括：
 * - 添加按钮/摇杆
 * - 保存/加载布局
 * - 摇杆模式设置
 * - 重置为默认布局
 * 
 * 供 GameControlEditorManager 和 ControlEditorActivity 共同使用
 */
public class ControlEditorOperations {
    private static final String TAG = "ControlEditorOperations";
    
    /**
     * 添加按钮到配置
     * 
     * @param context 上下文
     * @param config 控件配置
     * @param screenWidth 屏幕宽度
     * @param screenHeight 屏幕高度
     * @return 新创建的按钮数据
     */
    public static ControlData addButton(Context context, ControlConfig config, int screenWidth, int screenHeight) {
        if (config == null || config.controls == null) {
            return null;
        }
        
        ControlData button = new ControlData(context.getString(R.string.editor_default_button_name), ControlData.TYPE_BUTTON);
        button.x = screenWidth / 2f;
        button.y = screenHeight / 2f;
        button.width = 100;
        button.height = 100;
        button.opacity = 0.7f;
        button.visible = true;
        button.keycode = 62; // Space键
        
        config.controls.add(button);
        return button;
    }
    
    /**
     * 添加摇杆到配置
     * 
     * @param config 控件配置
     * @param screenWidth 屏幕宽度
     * @param screenHeight 屏幕高度
     * @return 新创建的摇杆数据
     */
    public static ControlData addJoystick(ControlConfig config, int screenWidth, int screenHeight) {
        if (config == null || config.controls == null) {
            return null;
        }
        
        ControlData joystick = ControlData.createDefaultJoystick();
        joystick.x = screenWidth / 2f;
        joystick.y = screenHeight / 2f;
        
        config.controls.add(joystick);
        return joystick;
    }
    
    /**
     * 添加文本控件到配置
     * 
     * @param context 上下文
     * @param config 控件配置
     * @param screenWidth 屏幕宽度
     * @param screenHeight 屏幕高度
     * @return 新创建的文本控件数据
     */
    public static ControlData addText(Context context, ControlConfig config, int screenWidth, int screenHeight) {
        if (config == null || config.controls == null) {
            return null;
        }
        
        String defaultTextName = context.getString(R.string.editor_default_text_name);
        ControlData text = new ControlData(defaultTextName, ControlData.TYPE_TEXT);
        text.x = screenWidth / 2f;
        text.y = screenHeight / 2f;
        text.width = 150;
        text.height = 150;
        text.opacity = 0.7f;
        text.bgColor = 0xFF808080; // 灰色背景（更清晰可见）
        text.visible = true;
        text.shape = ControlData.SHAPE_RECTANGLE; // 默认方形
        text.displayText = defaultTextName; // 默认文本
        text.keycode = ControlData.SDL_SCANCODE_UNKNOWN; // 文本控件不支持按键映射
        
        config.controls.add(text);
        return text;
    }
    
    /**
     * 显示摇杆模式批量设置对话框
     * 
     * @param context 上下文
     * @param config 控件配置
     * @param onLayoutUpdated 布局更新回调
     */
    public static void showJoystickModeDialog(Context context, ControlConfig config, 
                                            Runnable onLayoutUpdated) {
        if (config == null || config.controls == null) return;
        
        // 统计当前布局中的摇杆数量
        int joystickCount = 0;
        for (ControlData control : config.controls) {
            if (control.type == ControlData.TYPE_JOYSTICK) {
                joystickCount++;
            }
        }
        
        if (joystickCount == 0) {
            Toast.makeText(context, context.getString(R.string.editor_no_joystick), Toast.LENGTH_SHORT).show();
            return;
        }
        
        final String[] modes = {
            context.getString(R.string.editor_joystick_mode_keyboard),
            context.getString(R.string.editor_joystick_mode_mouse),
            context.getString(R.string.editor_joystick_mode_sdl)
        };
        
        new AlertDialog.Builder(context)
            .setTitle(context.getString(R.string.editor_joystick_mode_settings))
            .setMessage(context.getString(R.string.editor_joystick_mode_message, joystickCount))
            .setItems(modes, (dialog, which) -> {
                int newMode;
                String modeName;
                
                switch (which) {
                    case 0:
                        newMode = ControlData.JOYSTICK_MODE_KEYBOARD;
                        modeName = context.getString(R.string.editor_mode_keyboard_detailed);
                        break;
                    case 1:
                        newMode = ControlData.JOYSTICK_MODE_MOUSE;
                        modeName = context.getString(R.string.editor_mode_mouse_detailed);
                        break;
                    case 2:
                        newMode = ControlData.JOYSTICK_MODE_SDL_CONTROLLER;
                        modeName = context.getString(R.string.editor_mode_xbox_detailed);
                        break;
                    default:
                        return;
                }
                
                // 批量更新所有摇杆的模式
                int updatedCount = updateJoystickModes(context, config, newMode);
                
                // 通知布局已更新
                if (onLayoutUpdated != null) {
                    onLayoutUpdated.run();
                }
                
                Toast.makeText(context,
                    context.getString(R.string.editor_joysticks_set, updatedCount, modeName),
                    Toast.LENGTH_SHORT).show();
            })
            .setNegativeButton(context.getString(R.string.cancel), null)
            .show();
    }
    
    /**
     * 批量更新摇杆模式
     * 
     * @param context 上下文
     * @param config 控件配置
     * @param newMode 新模式
     * @return 更新的摇杆数量
     */
    private static int updateJoystickModes(Context context, ControlConfig config, int newMode) {
        int updatedCount = 0;
        String rightStick = context.getString(R.string.editor_stick_right);
        String leftStick = context.getString(R.string.editor_stick_left);
        
        for (ControlData control : config.controls) {
            if (control.type == ControlData.TYPE_JOYSTICK) {
                control.joystickMode = newMode;
                
                // 根据模式设置合适的默认值
                if (newMode == ControlData.JOYSTICK_MODE_KEYBOARD) {
                    // 键盘模式：确保有按键映射
                    if (control.joystickKeys == null || control.joystickKeys.length < 4) {
                        control.joystickKeys = new int[]{
                            ControlData.SDL_SCANCODE_W,  // up
                            ControlData.SDL_SCANCODE_D,  // right
                            ControlData.SDL_SCANCODE_S,  // down
                            ControlData.SDL_SCANCODE_A   // left
                        };
                    }
                } else if (newMode == ControlData.JOYSTICK_MODE_MOUSE) {
                    // 鼠标模式：清除按键映射
                    control.joystickKeys = null;
                } else {
                    // SDL控制器模式：清除按键映射，设置默认为左摇杆
                    control.joystickKeys = null;
                    if (control.name != null && control.name.contains(rightStick)) {
                        control.xboxUseRightStick = true;
                    } else if (control.name != null && control.name.contains(leftStick)) {
                        control.xboxUseRightStick = false;
                    }
                    // do nothing for others
                }
                updatedCount++;
            }
        }
        
        return updatedCount;
    }
    
    /**
     * 保存布局到 ControlLayoutManager
     * 
     * @param context 上下文
     * @param config 控件配置
     * @param layoutName 布局名称
     * @return 是否保存成功
     */
    public static boolean saveLayout(Context context, ControlConfig config, String layoutName) {
        if (config == null) {
            Toast.makeText(context, context.getString(R.string.editor_no_layout_to_save), Toast.LENGTH_SHORT).show();
            return false;
        }
        
        try {
            ControlLayoutManager manager = new ControlLayoutManager(context);
            com.app.ralaunch.model.ControlLayout layout = manager.getCurrentLayout();
            if (layout == null || !layout.getName().equals(layoutName)) {
                layout = new com.app.ralaunch.model.ControlLayout(layoutName != null ? layoutName : manager.getCurrentLayoutName());
            }
            
            // 清空现有元素
            layout.getElements().clear();
            
            // 转换 ControlData 为 ControlElement
            DisplayMetrics metrics = context.getResources().getDisplayMetrics();
            for (ControlData data : config.controls) {
                ControlElement element = ControlDataConverter.dataToElement(data, 
                    metrics.widthPixels, metrics.heightPixels);
                if (element != null) {
                    layout.addElement(element);
                }
            }
            
            // 保存到管理器
            manager.saveLayout(layout);
            
            Toast.makeText(context, context.getString(R.string.editor_layout_saved), Toast.LENGTH_SHORT).show();
            return true;
        } catch (Exception e) {
            AppLogger.error(TAG, "Failed to save layout", e);
            Toast.makeText(context, context.getString(R.string.editor_save_failed, e.getMessage()), Toast.LENGTH_SHORT).show();
            return false;
        }
    }
    
    /**
     * 从 ControlLayoutManager 加载布局
     * 
     * @param context 上下文
     * @param layoutName 布局名称（null 表示使用当前布局）
     * @return 加载的配置，如果失败返回 null
     */
    public static ControlConfig loadLayout(Context context, String layoutName) {
        try {
            ControlLayoutManager manager = new ControlLayoutManager(context);
            com.app.ralaunch.model.ControlLayout layout;
            
            if (layoutName != null) {
                // 查找指定名称的布局
                layout = null;
                for (com.app.ralaunch.model.ControlLayout l : manager.getLayouts()) {
                    if (l.getName().equals(layoutName)) {
                        layout = l;
                        break;
                    }
                }
            } else {
                // 使用当前布局
                layout = manager.getCurrentLayout();
            }
            
            if (layout == null || layout.getElements().isEmpty()) {
                return null;
            }
            
            // 转换 ControlElement 列表为 ControlConfig
            ControlConfig config = new ControlConfig();
            config.name = layout.getName();
            config.version = 1;
            config.controls = new java.util.ArrayList<>();
            
            DisplayMetrics metrics = context.getResources().getDisplayMetrics();
            for (ControlElement element : layout.getElements()) {
                ControlData data = ControlDataConverter.elementToData(element, 
                    metrics.widthPixels, metrics.heightPixels);
                if (data != null) {
                    config.controls.add(data);
                }
            }
            
            return config;
        } catch (Exception e) {
            AppLogger.error(TAG, "Failed to load layout", e);
            return null;
        }
    }
    
    /**
     * 重置为默认布局
     * 
     * @param context 上下文
     * @param controlLayout ControlLayout 实例（用于加载默认布局）
     * @param onResetComplete 重置完成回调
     */
    public static void resetToDefaultLayout(Context context, com.app.ralaunch.controls.ControlLayout controlLayout, 
                                           Runnable onResetComplete) {
        new AlertDialog.Builder(context)
            .setTitle(context.getString(R.string.editor_reset_default))
            .setMessage(context.getString(R.string.editor_reset_confirm))
            .setPositiveButton(R.string.game_menu_yes, (dialog, which) -> {
                if (controlLayout != null) {
                    controlLayout.loadDefaultLayout();
                    Toast.makeText(context, context.getString(R.string.editor_reset_complete), Toast.LENGTH_SHORT).show();
                }
                if (onResetComplete != null) {
                    onResetComplete.run();
                }
            })
            .setNegativeButton(R.string.game_menu_no, null)
            .show();
    }
}

File: app/src/main/java/com/app/ralaunch/controls/editors/ControlEditorOperations.kt
package com.app.ralaunch.controls.editors

import android.app.AlertDialog
import android.content.Context
import android.content.DialogInterface
import android.widget.Toast
import com.app.ralaunch.R
import com.app.ralaunch.controls.data.ControlData
import com.app.ralaunch.controls.packs.ControlPackManager
import org.koin.java.KoinJavaComponent
import com.app.ralaunch.controls.data.ControlData.Joystick
import com.app.ralaunch.controls.data.ControlData.KeyCode
import com.app.ralaunch.controls.data.ControlData.TouchPad
import com.app.ralaunch.controls.packs.ControlLayout

/**
 * 控件编辑器操作类
 *
 * 统一管理控件编辑器的共同操作，包括：
 * - 添加按钮/摇杆/触摸板/文本
 * - 保存/加载布局
 * - 摇杆模式设置
 */
object ControlEditorOperations {
    private val packManager: ControlPackManager by lazy {
        KoinJavaComponent.get(ControlPackManager::class.java)
    }
    
    val context: Context
        get() = KoinJavaComponent.get(android.content.Context::class.java)

    /**
     * 添加按钮到配置
     */
    fun addButton(layout: ControlLayout): ControlData.Button {
        val button = ControlData.Button()
        button.name = context.getString(R.string.editor_default_button_name)
        button.x = 0.5f
        button.y = 0.5f
        button.width = 0.15f
        button.height = 0.15f
        button.opacity = 0.5f
        button.isVisible = true
        button.keycode = KeyCode.KEYBOARD_SPACE

        layout.controls.add(button)
        return button
    }

    /**
     * 添加摇杆到配置
     */
    fun addJoystick(
        layout: ControlLayout, joystickMode: Joystick.Mode, isRightStick: Boolean
    ): Joystick {
        val joystick = Joystick()
        joystick.name = if (isRightStick) context.getString(R.string.joystick_right) else context.getString(R.string.joystick_left)
        joystick.mode = joystickMode
        joystick.isRightStick = isRightStick

        // 设置位置
        if (isRightStick) {
            joystick.x = 0.65f
        } else {
            joystick.x = 0.15f
        }
        joystick.y = 0.5f

        // 设置尺寸和样式
        joystick.width = 0.45f
        joystick.height = 0.45f
        joystick.opacity = 0.5f
        joystick.bgColor = -0xb5b5b6
        joystick.strokeColor = 0x66FFFFFF
        joystick.strokeWidth = 2f
        joystick.cornerRadius = 225f
        joystick.stickOpacity = 0.8f
        joystick.stickKnobSize = 0.35f

        // 设置按键映射（仅键盘模式需要）
        if (joystickMode == Joystick.Mode.KEYBOARD) {
            joystick.joystickKeys = arrayOf(
                KeyCode.KEYBOARD_W, KeyCode.KEYBOARD_D, KeyCode.KEYBOARD_S, KeyCode.KEYBOARD_A
            )
        }

        layout.controls.add(joystick)
        return joystick
    }

    /**
     * 添加触控板到配置
     */
    fun addTouchPad(layout: ControlLayout): TouchPad {
        val touchpad = TouchPad()
        touchpad.name = context.getString(R.string.editor_default_touchpad_name)
        touchpad.x = 0.5f
        touchpad.y = 0.5f
        touchpad.width = 0.45f
        touchpad.height = 0.45f
        touchpad.opacity = 0.5f
        touchpad.cornerRadius = 22.0f
        touchpad.isVisible = true

        layout.controls.add(touchpad)
        return touchpad
    }

    /**
     * 添加鼠标滚轮到配置
     */
    fun addMouseWheel(layout: ControlLayout): ControlData.MouseWheel {
        val mouseWheel = ControlData.MouseWheel()
        mouseWheel.name = context.getString(R.string.editor_default_mousewheel_name)
        mouseWheel.x = 0.5f
        mouseWheel.y = 0.5f
        mouseWheel.width = 0.2f
        mouseWheel.height = 0.3f
        mouseWheel.isSizeRatioLocked = false
        mouseWheel.opacity = 0.5f
        mouseWheel.cornerRadius = 15.0f
        mouseWheel.isVisible = true

        layout.controls.add(mouseWheel)
        return mouseWheel
    }

    /**
     * 添加文本控件到配置
     */
    fun addText(layout: ControlLayout): ControlData.Text {
        val defaultTextName = context.getString(R.string.editor_default_text_name)
        val text = ControlData.Text()
        text.name = defaultTextName
        text.x = 0.5f
        text.y = 0.5f
        text.width = 0.15f
        text.height = 0.15f
        text.opacity = 0.5f
        text.bgColor = -0x7f7f80
        text.isVisible = true
        text.shape = ControlData.Text.Shape.RECTANGLE
        text.displayText = defaultTextName

        layout.controls.add(text)
        return text
    }

    /**
     * 显示摇杆模式批量设置对话框
     */
    fun showJoystickModeDialog(
        context: Context, layout: ControlLayout?, onLayoutUpdated: Runnable?
    ) {
        if (layout == null) return

        // 统计当前布局中的摇杆数量
        var joystickCount = 0
        for (control in layout.controls) {
            if (control is Joystick) {
                joystickCount++
            }
        }

        if (joystickCount == 0) {
            Toast.makeText(
                context, context.getString(R.string.editor_no_joystick), Toast.LENGTH_SHORT
            ).show()
            return
        }

        val modes = arrayOf<String?>(
            context.getString(R.string.editor_joystick_mode_keyboard),
            context.getString(R.string.editor_joystick_mode_mouse),
            context.getString(R.string.editor_joystick_mode_sdl)
        )

        AlertDialog.Builder(context)
            .setTitle(context.getString(R.string.editor_joystick_mode_settings))
            .setMessage(context.getString(R.string.editor_joystick_mode_message, joystickCount))
            .setItems(
                modes
            ) { dialog: DialogInterface?, which: Int ->
                val newMode: Joystick.Mode?
                val modeName: String?

                when (which) {
                    0 -> {
                        newMode = Joystick.Mode.KEYBOARD
                        modeName = context.getString(R.string.editor_mode_keyboard_detailed)
                    }

                    1 -> {
                        newMode = Joystick.Mode.MOUSE
                        modeName = context.getString(R.string.editor_mode_mouse_detailed)
                    }

                    2 -> {
                        newMode = Joystick.Mode.GAMEPAD
                        modeName = context.getString(R.string.editor_mode_xbox_detailed)
                    }

                    else -> return@setItems
                }

                // 批量更新所有摇杆的模式
                val updatedCount = updateJoystickModes(layout, newMode)

                // 通知布局已更新
                onLayoutUpdated?.run()
                Toast.makeText(
                    context,
                    context.getString(R.string.editor_joysticks_set, updatedCount, modeName),
                    Toast.LENGTH_SHORT
                ).show()
            }.setNegativeButton(context.getString(R.string.cancel), null).show()
    }

    /**
     * 批量更新摇杆模式
     */
    private fun updateJoystickModes(layout: ControlLayout, newMode: Joystick.Mode): Int {
        var updatedCount = 0

        layout.controls.mapNotNull { it as? Joystick }.forEach { joystick ->
            joystick.mode = newMode
            updatedCount++
        }

        return updatedCount
    }

    /**
     * 保存布局到 ControlPackManager
     */
    fun saveLayout(context: Context, layout: ControlLayout?): Boolean {
        if (layout == null) {
            Toast.makeText(
                context, context.getString(R.string.editor_no_layout_to_save), Toast.LENGTH_SHORT
            ).show()
            return false
        }

        try {
            val packId = packManager.getSelectedPackId()
            
            if (packId != null) {
                packManager.savePackLayout(packId, layout)
                Toast.makeText(
                    context, context.getString(R.string.editor_layout_saved), Toast.LENGTH_SHORT
                ).show()
                return true
            } else {
                Toast.makeText(
                    context, context.getString(R.string.editor_no_layout_to_save), Toast.LENGTH_SHORT
                ).show()
                return false
            }
        } catch (e: Exception) {
            Toast.makeText(
                context,
                context.getString(R.string.editor_save_failed, e.message),
                Toast.LENGTH_SHORT
            ).show()
            return false
        }
    }

    /**
     * 从 ControlPackManager 加载布局
     */
    fun loadLayout(packId: String?): ControlLayout? {
        return try {
            if (packId != null) {
                packManager.getPackLayout(packId)
            } else {
                packManager.getCurrentLayout()
            }
        } catch (_: Exception) {
            null
        }
    }
}
